### DISCUSSION TITLE
How can I set the ReactCompareSlider to maintain aspect ratio, with no cropping, and automatically resize to the available space?

### DISCUSSION THREAD

**Discussion by tpdickson** [nerdyman/react-compare-slider#145] [Q&A]
I'm having trouble getting the ReactCompareSlider to behave how I want it to with regards to taking up the available space, and maintaining the aspect ratio.

For example, if I have an img:

```tsx
<img
  src={originalImage}
  alt=""
  className="size-full object-contain"
/>
```

with Tailwind CSS classes `size-full` and `object-contain`, it will take up all of the available space in the parent container, up to the point that either the width or height reaches the edges, then it will automatically scale the other dimension to maintain the correct aspect ratio:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/da8175af-9f7f-4af9-a582-fc309942cbfb)

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/18e90355-14c3-4290-bf51-7e613443f5bc)

(in this case, the parent container takes up the entire screen, and is an "overlay")

However, when I try to get the same behavior with ReactCompareSlider, I get all sorts of funky results:

Simply putting `size-full` and `object-contain` on the ReactCompareSlider:

```tsx
<ReactCompareSlider
  className="size-full object-contain"
  itemOne={<img src={originalImage} alt="" className="" />}
  itemTwo={<img src={originalImage} alt="" className="" />}
/>
```

(I'm using the same image in itemOne and itemTwo for simple testing right now.)

Here's what it looks like:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/b6430114-950e-4f31-b7fd-9f893284408e)

Then if I also put `size-full` and `object-contain` on the items' img elements:

```tsx
<ReactCompareSlider
  className="size-full object-contain"
  itemOne={<img src={originalImage} alt="" className="size-full object-contain" />}
  itemTwo={<img src={originalImage} alt="" className="size-full object-contain" />}
/>
```

It looks good initially:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/e049c31b-f44e-4413-82cf-e8bb070ba99d)

But the actual slider part of it ends up overlaying the entire screen, since it is "detached" from the aspect ratio of the item components:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/2ca64919-54b9-402d-a86d-199d11996290)

A side effect is that I can't click the buttons on the top right, since the slider takes up the entire screen:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/134bbeaf-83ce-44a6-a6a2-35660479c6e1)

---

How would I go about fixing this to get it to behave as I want it to?

**Comment by nerdyman** [comment]
@tpdickson  I'm not certain but think `size-full` on the slider component itself is the problem. It shouldn't need any custom classes to fit correctly as the slider always fits to the bounds of whatever is in `itemOne`.

It looks like `size-full` is setting everything to be full width and height but the image is set to contain so the slider is filling the container and the image is being scaled down.

**Reply by tpdickson** [reply]
Thanks for looking at this quickly.

Here's with the slider component css classes removed:

```tsx
<ReactCompareSlider
  itemOne={<img src={originalImage} alt="" className="size-full object-contain" />}
  itemTwo={<img src={originalImage} alt="" className="size-full object-contain" />}
/>
```

It doesn't scale up to the entire screen when the screen is larger than the base image resolution:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/0cb95b80-cb82-4e6b-a399-c5db861bd2c5)

It doesn't scale down when the screen is smaller vertically:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/11a60e10-4d93-449a-8b8a-1b459ee36327)
(it does, however, correctly keep the slider within the bounds of the image)

It *does* behave exactly how I want it to when the screen is smaller horizontally than the image:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/9d0ac98a-1d1a-4ae3-b5d7-90cd2a3d1549)

---

So the next attempt is to change `size-full` (height: 100%, width: 100%) to `h-screen` and `w-screen` (height: 100vh, width: 100vw):

```tsx
<ReactCompareSlider
  itemOne={<img src={originalImage} alt="" className="h-screen w-screen object-contain" />}
  itemTwo={<img src={originalImage} alt="" className="h-screen w-screen object-contain" />}
/>
```

This causes the ReactCompareSlider to overflow the bounds of the image. The image, however, is doing exactly what I want it to do. It correctly scales down when the width of the screen is squished:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/b0cc9453-57f4-4256-a63f-dffb10cee5e4)

Also when the height of the screen is squished:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/ee055c54-64e0-4e90-a6f5-79392a08cb9d)

When the screen is larger than the image, it also correctly scales up:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/4d298396-b125-4706-a7a0-e43dded5f3e0)

It seems like the `object-contain` (object-fit: contain) is being applied to the images, but not doing anything to restrict the bounds of the parent ReactCompareSlider.

---

Sorry if I'm missing something fundamental to how this works, I'm very new to web front end.

**Reply by nerdyman** [reply]
@tpdickson No worries, it's a bit of a tricky issue. My first comment about the slider fitting the bounds won't work when `object-fit` is used because the image width and height doesn't change, the browser scales down the image to fit but the actual width and height remain the same. There are [more deets on MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Replaced_element#controlling_object_position_within_the_content_box).

I'll look at getting a repro and post a link here, let me know if you find a solution in the meantime!

**Comment by nerdyman** [comment]
@tpdickson I think this does what you need, it fills the width and height up to the intrinsic size of the image and scales it down without clipping it when the bounds shrink.

https://stackblitz.com/edit/vitejs-vite-di5xps?file=src%2FApp.tsx&terminal=dev

```tsx
<ReactCompareSlider
    style={{ maxHeight: '100%' }}
    itemOne={
      <ReactCompareSliderImage
        src="https://dummyimage.com/640x360/aaa/eee?text=One"
        alt="one"
        style={{ objectFit: 'contain', height: 'auto', maxHeight: '100%' }}
      />
    }
    itemTwo={
      <ReactCompareSliderImage
        src="https://dummyimage.com/640x360/eee/aaa?text=Two"
        alt="Two"
        style={{ objectFit: 'contain', height: 'auto', maxHeight: '100%' }}
      />
    }
  />
```

**Reply by tpdickson** [reply]
With that code, when the height is smaller than the image, the slider is wider than the images:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/a145600a-cf01-4bea-976a-82fe6d47c8ec)

When the window is larger than the image, the image doesn't upscale to the size of the window:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/724d1ca1-eb93-41be-9b85-59036477816d)

---

I also noticed that there is quite different behavior between using Tailwind CSS's `className` and tailwind css, compared to using `style` with normal css.

There is also different behavior between using `img` and `ReactCompareSliderImage` in some cases, which I guess is expected, otherwise `ReactCompareSliderImage` wouldn't have much use.

---

This gets close:

```tsx
  <ReactCompareSlider
    itemOne={
      <ReactCompareSliderImage
        src="https://dummyimage.com/640x360/aaa/eee?text=One"
        alt="one"
        style={{ objectFit: "contain", height: "100vh", maxHeight: "100vw" }}
      />
    }
    itemTwo={
      <ReactCompareSliderImage
        src="https://dummyimage.com/640x360/eee/aaa?text=Two"
        alt="two"
        style={{ objectFit: "contain", height: "100vh", maxHeight: "100vw" }}
      />
    }
  />
```

It fills up the screen properly when the screen is larger than the image:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/a7aa96b6-802c-4a82-8616-624c07aafb7e)

It scales down properly when vertically squished:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/5dfa79e4-eccd-4f03-939e-49960d0bebe0)

However, when horizontally squished, the compare slider is taller than the image:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/8c5e81a2-b1de-48e4-876e-4fddca9a2b60)

**Reply by nerdyman** [reply]
The StackBlitz demo does what you need so the issue is most likely due to the styles on the parent element rendering the slider. 

## Doesn't exceed image bounds when screen is bigger than image

![image](https://github.com/nerdyman/react-compare-slider/assets/4908432/853477d8-baa8-43d2-bce8-e2e02a865959)

## Scales horizontally when screen is smaller than image

![image](https://github.com/nerdyman/react-compare-slider/assets/4908432/5b9bcf8c-8af3-4097-bc5f-66212071e99f)


## Scales Vertically when screen is smaller than image

![image](https://github.com/nerdyman/react-compare-slider/assets/4908432/681cec20-b448-4ab3-85a4-4e6bbab4601f)


The demo applies some flexbox styles to the parent so it's worth trying the same setup as the demo and making your customisations from there to see what's breaking things. If you can provide a repro on StackBlitz I'll be able to take a look.

The slider has default inline styles so I'd use inline styles instead of Tailwind if possible. Otherwise you'll have to `!important` your Tailwind styles to override the slider ones. Also, here's the [docs for `ReactCompareSliderImage`](https://react-compare-slider.vercel.app/?path=/docs/docs-images--docs#reactcomparesliderimage) detailing what it does.

**Reply by tpdickson** [reply]
In the stackblitz you linked: https://stackblitz.com/edit/vitejs-vite-di5xps?file=src%2FApp.tsx&terminal=dev

The slider exceeds the bounds of the image when vertically compressed:
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/5127bf93-c622-47e5-8c98-9d33c006caf6)

And it *doesn't* scale up the images to fill the available space when the viewport is larger than the images (I want it to scale up):
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/bed38cac-ec57-4062-936b-230b6108fdfa)

That's the same behavior as when I tried the code in the page I'm working on.

I'll play around with the parent container though to see if I can get something better.

**Reply by tpdickson** [reply]
I think the issue is that in this code:

```tsx
<ReactCompareSliderImage
  src="https://dummyimage.com/640x360/aaa/eee?text=One"
  alt="one"
  style={{ objectFit: "contain", height: "100vh", maxHeight: "100vw" }}
/>
```

The `ReactCompareSliderImage` is actually larger than the displayed image:

![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/beb5edc5-d6ad-496b-b8d4-75c339b9466b)
![image](https://github.com/nerdyman/react-compare-slider/assets/3342510/7a82c2e7-5b92-4703-b01c-ccdd6aa0aef9)

I need to find a way to get the same behavior on the image (scale up to the size of the viewport, while still maintaining the correct aspect ratio that `objectFit: 'contain'` provides), without having the img component be larger than the displayed image itself.

**Reply by nerdyman** [reply]
@tpdickson Gotcha 👌 This is a lot trickier than I thought. I'll try and get a working example.

**Reply by tpdickson** [reply]
Yea to me at least, this has been a doozy. I've been playing around with the styles, trying to get the same behavior in a simple:

```tsx
<div
  style={{ }}
>
  <img
    src="https://dummyimage.com/640x360/aaa/eee?text=One"
    alt="one"
    style={{ }}
  />
</div>
```

and have had no luck. Always at least one undesirable behavior no matter what I've tried. So it's not necessarily a problem with ReactCompareSlider, more of a problem with getting the CSS correct (if it's even possible).

I'm leaning more and more towards the thought that I'll have to ditch having the images scale up to the screen size.

Thanks for the help again though, I really do appreciate it.

**Reply by tpdickson** [reply]
I asked on the Tailwind CSS discord about getting the behavior of `width: 100vw, height: 100vh, objectFit: contain`, but without the `img` element being larger than the displayed image, and someone who works on Tailwind replied that I should look into using JavaScript to get it to work. It's very likely not doable with just CSS.

**Reply by nerdyman** [reply]
@tpdickson If you know the aspect ratio of the image you can use the `aspect-ratio` CSS property to force the slider to keep those bounds. I've updated the demo, _hopefully_ it works 🙏 

If you don't know it initially you can use JS to get the [natural width and height of the image](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/naturalHeight) to determine the aspect ratio.

**Reply by tpdickson** [reply]
Unfortunately there's no upscaling on that. I tried a useEffect approach and calculating the image size based on the window size. It works for properly resizing the image, now I need to figure out how to get the slider to be properly contained and centered:

```tsx
import {
  ReactCompareSlider,
  ReactCompareSliderImage,
} from 'react-compare-slider';
import './App.css';
import { useState, useEffect } from 'react';

function App() {
  const [imageSize, setImageSize] = useState({
    width: 0,
    height: 0,
  });
  const [windowSize, setWindowSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  useEffect(() => {
    const handleResize = () => {
      setWindowSize({
        width: window.innerWidth,
        height: window.innerHeight,
      });
    };

    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  const imageDisplaySize = calculateImageDisplaySize(
    windowSize.width,
    windowSize.height,
    imageSize.width,
    imageSize.height
  );

  return (
    <ReactCompareSlider
      style={{}}
      itemOne={
        <ReactCompareSliderImage
          src="https://dummyimage.com/640x360/aaa/eee?text=One"
          alt="One"
          style={{
            margin: 'auto',
            width: `${imageDisplaySize.width}px`,
            height: `${imageDisplaySize.height}px`,
          }}
          onLoad={(event: React.SyntheticEvent<HTMLImageElement>) => {
            setImageSize({
              width: event.currentTarget.naturalWidth,
              height: event.currentTarget.naturalHeight,
            });
          }}
        />
      }
      itemTwo={
        <ReactCompareSliderImage
          src="https://dummyimage.com/640x360/eee/aaa?text=Two"
          alt="Two"
          style={{
            margin: 'auto',
            width: `${imageDisplaySize.width}px`,
            height: `${imageDisplaySize.height}px`,
          }}
        />
      }
    />
  );
}

export default App;

function calculateImageDisplaySize(
  windowWidth: number,
  windowHeight: number,
  imageWidth: number,
  imageHeight: number
): { width: number; height: number } {
  // calculate the image display dimensions
  const windowAspectRatio = windowWidth / windowHeight;
  const imageAspectRatio = imageWidth / imageHeight;

  let newImageWidth, newImageHeight;

  if (imageAspectRatio > windowAspectRatio) {
    newImageWidth = windowWidth;
    newImageHeight = imageHeight * (windowWidth / imageWidth);
  } else if (imageAspectRatio < windowAspectRatio) {
    newImageWidth = imageWidth * (windowHeight / imageHeight);
    newImageHeight = windowHeight;
  } else {
    newImageWidth = windowWidth;
    newImageHeight = windowHeight;
  }

  return { width: newImageWidth, height: newImageHeight };
}
```

https://github.com/nerdyman/react-compare-slider/assets/3342510/25cb6f45-0dae-4c98-ac7c-6110402b4504

---

The `margin: 'auto'` on the images is just the current attempt at getting the images properly centered and the slider contained around them, it doesn't work though. I need to play around with it more.

**Reply by tpdickson** [reply]
I added a `div` around the slider and it seems to do everything I want it to do.

Here's the `div`:

```tsx
<div
  style={{
    display: 'flex',
    justifyContent: 'center',
  }}
>
  ...
</div>
```

And here's the full demo code:

```tsx
import {
  ReactCompareSlider,
  ReactCompareSliderImage,
} from 'react-compare-slider';
import './App.css';
import { useState, useEffect } from 'react';

function App() {
  const [imageSize, setImageSize] = useState({
    width: 0,
    height: 0,
  });
  const [windowSize, setWindowSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  useEffect(() => {
    const handleResize = () => {
      setWindowSize({
        width: window.innerWidth,
        height: window.innerHeight,
      });
    };

    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  const imageDisplaySize = calculateImageDisplaySize(
    windowSize.width,
    windowSize.height,
    imageSize.width,
    imageSize.height
  );

  return (
    <div // <----------------------------------------this div and its style are new
      style={{
        display: 'flex',
        justifyContent: 'center',
      }}
    >
      <ReactCompareSlider
        style={{}}
        itemOne={
          <ReactCompareSliderImage
            src="https://dummyimage.com/640x360/aaa/eee?text=One"
            alt="One"
            style={{
              width: `${imageDisplaySize.width}px`,
              height: `${imageDisplaySize.height}px`,
            }}
            onLoad={(event: React.SyntheticEvent<HTMLImageElement>) => {
              setImageSize({
                width: event.currentTarget.naturalWidth,
                height: event.currentTarget.naturalHeight,
              });
            }}
          />
        }
        itemTwo={
          <ReactCompareSliderImage
            src="https://dummyimage.com/640x360/eee/aaa?text=Two"
            alt="Two"
            style={{
              width: `${imageDisplaySize.width}px`,
              height: `${imageDisplaySize.height}px`,
            }}
          />
        }
      />
    </div>
  );
}

export default App;

function calculateImageDisplaySize(
  windowWidth: number,
  windowHeight: number,
  imageWidth: number,
  imageHeight: number
): { width: number; height: number } {
  const windowAspectRatio = windowWidth / windowHeight;
  const imageAspectRatio = imageWidth / imageHeight;

  let newImageWidth, newImageHeight;

  if (imageAspectRatio > windowAspectRatio) {
    newImageWidth = windowWidth;
    newImageHeight = imageHeight * (windowWidth / imageWidth);
  } else if (imageAspectRatio < windowAspectRatio) {
    newImageWidth = imageWidth * (windowHeight / imageHeight);
    newImageHeight = windowHeight;
  } else {
    newImageWidth = windowWidth;
    newImageHeight = windowHeight;
  }

  return { width: newImageWidth, height: newImageHeight };
}
```

---

I still wonder if there's a way to do it in pure CSS without any javascript/typescript, but... that's for another time.

Thanks a ton for the help. Narrowing down the problem made it quite a bit easier to come to a full solution, and also a simpler, CSS-only solution if I do decide to get rid of upscaling in the future.
